### ISSUE TITLE
OIDC token not automatically refreshed

### ISSUE THREAD

**Issue by Morriz** [kiali/kiali#5233]
Hi, it seems Kiali is not following the preferred oidc approach, where an expired token should result in a token refresh request with the authenticating party. If that WOULD be the case Kiali would have no dealings with session duration whatsoever imo.

To put it simply: imo Kiali should not do anything with session duration (`auth.openid.authentication_timeout`) when `auth.strategy: openid` is chosen, as it can rely on the oauth flow to invalidate itself, and a redirect to the IDP's login page is served.

**Comment by israel-hdez** [comment]
Yes, the token refresh is not implemeted in Kiali. People have asked for this in the past. However, we are looking for a community contribution to cover that part of the openid spec.

That said, I think most people simply reconfigure their IdP to use long lived tokens and have been happy with that. So, the perception is that the community is not that interested in the token refresh part.

> To put it simply: imo Kiali should not do anything with session duration (auth.openid.authentication_timeout) when auth.strategy: openid is chosen

It doesn't. Kiali keeps the session open until the `id_token` expires.

The `authentication_timeout` configuration is something else. Read the small explanation here: https://kiali.io/docs/configuration/authentication/openid/#configuring-authentication-timeout

**Comment by Morriz** [comment]
Ok, I see. But that is breaking the chain of validity: kiali might think the user is valid, but the IdP might think differently. It is preferred to have expire the key as fast as possible without it having perf impact (`60s` I see a lot). How much effort is it to add that flow?

**Comment by israel-hdez** [comment]
> kiali might think the user is valid, but the IdP might think differently

I do agree with that.

> How much effort is it to add that flow?

Speaking technically, I think it is in between a small-medium size task. I do not think it is complex if the spec is well understood.

Testing and reviewing the change is what usually takes more time, because of the needed knowledge about OIDC and the Kiali codebase related to it, and also because of the setup of a testing environment.

**Comment by Morriz** [comment]
so it would be least effort for a core maintainer. Is Kiali not funded or backed by VC?

**Comment by jshaughn** [comment]
Hi @Morriz. Kiali is an open source project, not really backed by VC.  The core contributors are primarily funded by Red Hat, and as such OpenShift authorization is the usual priority.  Other authentication schemes are offered, and worked on as time permits, to support the general Istio community.  We have received contributions in this area before (usually guided by @israel-hdez), and it is often due to a community member needing some capability or refinement, like in this case.  Left to the core team, no timetable can be given for this task.  Not because we wouldn't want to satisfy everyone, and not because it's not useful, but just because we're not a huge team and planning-wise we're booked for the foreseeable future.

**Comment by Morriz** [comment]
But I  would assume openshift auth is OIDC based. Isn't it? Then it would token refresh as well, no? Why not backport it here?

**Comment by jmazzitelli** [comment]
Within the Kiali Server, the "openshift" auth strategy is implemented differently than the "openid" auth strategy. OpenShift's default internal OAuth mechanism (that which the Kiali "openshift' auth strategy integrates with) is not OpenID.... https://docs.openshift.com/container-platform/4.10/authentication/configuring-internal-oauth.html#configuring-internal-oauth

**Comment by israel-hdez** [comment]
Right, as @jmazzitelli is saying,  OpenShift auth is based on OAuth rather than OIDC. Quite similar, but not the same. And FWIW, it also does not have token refresh.

If OpenShift would use OIDC, we would haven't had the need to implement two auth schemes.

Curiously, Kiali's OIDC implementation is more robust that OpenShift auth, because OIDC is using the `authorization code flow`, while OpenShift is using the less secure `implicit flow`. So, porting some code from OIDC auth to OpenShift auth is more important at the moment (related issue #4890). Even then, hardening OpenShift auth is not a priority at the moment.

**Comment by Morriz** [comment]
Got it! Thanks for the detailed feedback...much appreciated.
